plugins/buildingplan/buildingplan_cycle.cpp

```diff
@@ -228,84 +228,125 @@ static df::building * popInvalidTasks(color_ostream &out, Bucket &task_queue,
-        if (!itemPassesScreen(out, item))
-            continue;
-        for (auto bucket_it = buckets.begin(); bucket_it != buckets.end(); ) {
-            TRACE(cycle,out).print("scanning bucket: %s/%s\n",
-                    ENUM_KEY_STR(job_item_vector_id, vector_id).c_str(), bucket_it->first.c_str());
-            auto & task_queue = bucket_it->second;
-            auto bld = popInvalidTasks(out, task_queue, planned_buildings);
-            if (!bld) {
-                DEBUG(cycle,out).print("removing empty bucket: %s/%s; %zu bucket(s) left\n",
-                      ENUM_KEY_STR(job_item_vector_id, vector_id).c_str(),
-                      bucket_it->first.c_str(),
-                      buckets.size() - 1);
-                bucket_it = buckets.erase(bucket_it);
-                continue;
-            }
+
+    //  items we might want to attach (and their positions)
+    std::vector<df::item*> available;
+    for (auto item : item_vector) {
+        if (itemPassesScreen(out, item))
+            available.emplace_back(item);
+    }
+    std::vector<std::pair<df::coord, df::item*>> matching;
+    size_t num_matching = 0;
+
+    DEBUG(cycle,out).print("%zu items available for assignment\n", available.size());
+
+    for (auto bucket_it = buckets.begin(); bucket_it != buckets.end(); ) {
+
+        TRACE(cycle,out).print("scanning bucket: %s/%s\n",
+              ENUM_KEY_STR(job_item_vector_id, vector_id).c_str(), bucket_it->first.c_str());
+
+        auto & task_queue = bucket_it->second;
+        bool first_task = true;
+
+        while (auto bld = popInvalidTasks(out, task_queue, planned_buildings)){
             auto & task = task_queue.front();
+
             auto id = task.first;
             auto job = bld->jobs[0];
-            auto & jitems = job->job_items;
+            auto &jitems = job->job_items;
             const size_t num_filters = jitems.size();
-            auto filter_idx = task.second;
+            const int filter_idx = task.second;
             const int rev_filter_idx = num_filters - (filter_idx+1);
             auto &pb = planned_buildings.at(id);
-            if (matchesFilters(item, jitems[filter_idx], pb.heat_safety,
-                        pb.item_filters[rev_filter_idx], pb.specials)
-                    && Job::attachJobItem(job, item,
-                        df::job_item_ref::Hauled, filter_idx))
-            {
+
+            // first task of the bucket: filter/count available items
+            if (first_task) {
+                matching.clear();
+                for(auto item : available)
+                    if (!item->flags.bits.in_job &&
+                        matchesFilters(item,
+                                       jitems[filter_idx],
+                                       pb.heat_safety,
+                                       pb.item_filters[rev_filter_idx],
+                                       pb.specials))
+                        matching.emplace_back(item->pos,item);
```

Comment:
you should probably use `Items::getPosition(item)` instead of `item->pos` since the `pos` field may be garbage if the item is in a container

Reply:
Thanks, will do.